config/configgates/gates.go

```diff
@@ -0,0 +1,127 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package configgates
+
+import (
+	"fmt"
+	"sync"
+)
+
+// Gates is the interface used to query the status of feature gates. It can be
+// implemented by a read-only copy of the feature gate registry.
+type Gates interface {
+	IsEnabled(id string) bool
+}
+
+// Gate represents an individual feature that may be enabled or disabled based
+// on the lifecycle state of the feature and CLI flags specified by the user.
+type Gate struct {
+	ID          string
+	Description string
+	Enabled     bool
+}
+
+// Registry holds information about available feature gates and their enablement status.
+//
+// Components that wish to register a feature gate should do so using the global Registry
+// available from GetRegistry() in an init() function to ensure that the gate is registered
+// as soon as possible and thus available to be managed by configuration and CLI flags.
+type Registry struct {
+	sync.RWMutex
```

Comment:
Can we require that all registrations happen in main goroutine before the flags are queried (and before their values are loaded from the command line flag)? If we do this then the mutex and Frozen() is no longer needed.
I think we need that as a requirement anyway otherwise there is a danger of seeing different values of gates depending on when they are frozen/queried.

Reply:
I think we can require that by policy but I don't know that there are technical means of enforcing that policy available to us.  I think a statement that `Gate`s may only be added in `init()` functions and that the `Registry` is not thread-safe may be sufficient to enable removing the mutex.  `Frozen()` will be replaced with an `Apply(map[string]bool)` method that applies a configuration to a frozen copy of the `Registry`.  This is idempotent and safe against threading issues as long as the configuration map is.